major refactoring (#8)
gitsbi/unlib: README.md

# unlib
A minimal C++14 unit library

> Unlib is a minimal, C++14-compatible [SI unit](https://en.wikipedia.org/wiki/International_System_of_Units) library of arithmetic types, providing [quantities](#quantities) that mainly feature the seven [basic units](#units) (e.g., Watt) and derived units (e.g., Ws), the [physical dimensions](#dimensions) (e.g. power) built upon them, and the [scaling](#scales) (e.g., _kilo_) of the units. Quantities hold values in your code and mostly behave like C++' built-in arithmetic types: they can be added, multiplied, compared, etc. Quantities representing incompatible physical dimensions are incompatible to each other, and the compiler keeps track of the correct types of the results of arithmetic operations. For example, multiplying a quantity of the unit _kW_ (_power_) with a quantity of the unit _h_ (_time_) will result in a quantity of the unit _kWh_ _(energy_).

Reviewer: ```suggestion
Unlib is a minimal, C++14-compatible [SI unit](https://en.wikipedia.org/wiki/International_System_of_Units) library of arithmetic types, providing [quantities](#quantities) that mainly feature the seven [base units](#units) (e.g., kilogram, metre) and derived units (e.g., Ws), the [physical dimensions](#dimensions) (e.g. power) built upon them, and the [scaling](#scales) (e.g., _kilo_) of the units. Quantities hold values in your code and mostly behave like C++' built-in arithmetic types: they can be added, multiplied, compared, etc. Quantities representing incompatible physical dimensions are incompatible to each other, and the compiler keeps track of the correct types of the results of arithmetic operations. For example, multiplying a quantity of the unit _kW_ (_power_) with a quantity of the unit _h_ (_time_) will result in a quantity of the unit _kWh_ _(energy_).
```

Author: ```suggestion
Unlib is a minimal, C++14-compatible [SI unit](https://en.wikipedia.org/wiki/International_System_of_Units) library of arithmetic types, providing [quantities](#quantities) that mainly feature the seven [base units](#units) (e.g., kilogram, meter) and derived units (e.g., Ws), the [physical dimensions](#dimensions) (e.g. power) built upon them, and the [scaling](#scales) (e.g., _kilo_) of the units. Quantities hold values in your code and mostly behave like C++' built-in arithmetic types: they can be added, multiplied, compared, etc. Quantities representing incompatible physical dimensions are incompatible to each other, and the compiler keeps track of the correct types of the results of arithmetic operations. For example, multiplying a quantity of the unit _kW_ (_power_) with a quantity of the unit _h_ (_time_) will result in a quantity of the unit _kWh_ _(energy_).
```

The library tries to protect you from Murphy's doings and therefore errs on the side of requiring you to be explicit, rather than on the side of conveniences, where a design decision between these two had to be made.

## Units

> The SI unit system is built on the [seven basic units](https://en.wikipedia.org/wiki/SI_base_unit):

Reviewer: ```suggestion
The SI unit system is built on the [seven base units](https://en.wikipedia.org/wiki/SI_base_unit):
```

Author: 👍🏻

```

Since the library's way of telling these exponents from each other is their order, they must always be passed in the correct order, which is an invitation to silly errors. As a human, in order to even stand a chance to write this without making an error, you would have to write it this way:

```cpp
using my_dimension = unlib::dimension< /* time               */ unlib::ratio_t<1>  
                                     , /* mass               */ unlib::ratio_t<0> 
                                     , /* length             */ unlib::ratio_t<0> 
                                     , /* current            */ unlib::ratio_t<1> 
                                     , /* temperature        */ unlib::ratio_t<0> 
                                     , /* substance amount   */ unlib::ratio_t<0> 
                                     , /* luminous intensity */ unlib::ratio_t<0> >;
```

However, dimensions can, and should, be built in different, less error-prone, ways. For example, the `dimension_t` meta function creates dimensions from the exponents of the seven basic units, passed in any order. Using it, electrical charge can be defined this way:

```c++

> using electrical_charge = unlib::dimension_t<unlib::current, unlib::time>;

Reviewer: This would be much clearer written as
`using electrical_charge = unlib::mul_dimension_t<unlib::current, unlib::time>;`

IMHO, the "promoted" way to derive dimensions should be in terms of operations, not using `unlib::dimension_t` directly, as in this example. I would even go so far as to not expose this as user interface. This would also open up the opportunity to simplify the code by removing the metafunctions to handle "unsorted unit lists".

Author: OK. I will look into this and maybe remove `dimension_t` from the interface altogether.

```c++
using mass_per_length = unlib::dimension_t< unlib::mass
                                          , unlib::reciprocal_dimension_t<unlib::length> >;
```

But this might be even easier to understand if you use the `div_dimension_t` meta function: 

```c++
using mass_per_length = unlib::div_dimension_t< unlib::mass
                                              , unlib::length >
```

(_Note: The type of a quantity defines the nested type `dimension_type`, which could also be used to define this dimension:_

```c++
using namespace unlib::literals;
auto one_kg_per_meter = 1_kg/1_m;
using mass_per_length = decltype(one_kg_per_meter)::dimension_type;
```

_Quantities are explained [further down](#quantities)._) 

## Scales

Units are scaled, where scales are rational numbers, represented by `std::ratio`. Due to the limitations of 64 bit integer arithmetic with `std::ratio`, of the [standard scales provided by the SI system](https://en.wikipedia.org/wiki/International_System_of_Units#Prefixes) this library only covers femto–exa. They are named like `unlib::femto_scaling`, `unlib::pico_scaling` etc., but they are simply aliases for `std::femto`, `std::pico`, etc. 

> Besides those, noteworthy pre-defined scales are `no_scaling`, which denotes and inscaled quantity, as well as `minute_scaling`, `hour_scaling`, `day_scaling`, and `week_scaling`, which can be used to properly scale time quantities according to the Babylonian scaling system we use for time units.

Reviewer: ```suggestion
Besides those, noteworthy pre-defined scales are `no_scaling`, which denotes an unscaled quantity, as well as `minute_scaling`, `hour_scaling`, `day_scaling`, and `week_scaling`, which can be used to properly scale time quantities according to the Babylonian scaling system we use for time units.
```

Author: 👍🏻

A quantity combines a [unit](#units) (_mass_), a [scale](#scales) (_kilo_), a specific [value type](#values) (_double_), and an optional [tag](#tags) (_reactive_): 

```c++

> using kilogram = unlib::quantity<unlib::mass,unlib::kilo,double>;

Reviewer: this is wrong, the base unit of mass is already kilogram, what you define here is megagram. Also, `unlib::kilo` is an alias for `scale_by`, what you want here is the scaling explicitly.
```suggestion
using kilogram = unlib::quantity<unlib::mass,unlib::no_scaling,double>;
```

Author: Ugh. OK, as we talked about, `kilogram<>` will again become `kilo<gram<>>`, which would make `unlib::kilo_scaling` the correct identifier here.

However, since this will be done on a different branch, I just avoided the issue by using kilometer as an example here:

```suggestion
using kilometer = unlib::quantity<unlib::length,unlib::kilo_scaling,double>;
```

> using     watt = div_quantity_t<joule<V>,second<V>>;

Reviewer: ```suggestion
template<typename V>
using     watt = div_quantity_t<joule<V>,second<V>>;
```

Author: Oops. To make the examples simpler, I have always just used a fixed value type.

```

_(Note: All of unlib's predefined quantity types, including _kilogram_, are templates that take the value type – here fixed to **`double`** – as a template parameter.)_ 

A quantity can be created from a value type explicitly. It cannot be implicitly created from a value type, though: 

```c++
void f(unlib::quantity<unlib::mass,unlib::kilo,double>);

kilogram kg{42}; // fine
f(kg);           // fine, too
f(42);           // won't compile
f(kilogram{42}); // also fine

using namespace unlib::literals; // needed for unit suffixes

> f(42_kg);        // also works

Reviewer: ```suggestion
f(42.0_kg);        // also works
f(23_kg);           // does not work
```
the integer literal will create an integer-valued quantity, which is not accepted by `f`, as there is no implicit value type cast.

Author: I would not bring the value type into this, as that is dealt with only later. But that was a good catch, I am using a double literal now.

```

A quantity also cannot be used where its value type is needed. If you need to pass a quantity to an API that expects one of the built-in value types instead, you have to explicitly convert it:

```c++
void f(double kg);

f(some_mass.get());                             // returns underlying value, whatever scaling

> f(some_mass.get_scaled<unlib::kilo_scaling>()); // will always return kg 

Reviewer: same mistake again, this will actually return megagram.

looking at this example and the problem that `unlib::quantity<unlib::mass,unlib::kilo,double>` is actually "Megagram", I'm not convinced that making kilogram the "unscaled" base unit is a good choice.

It is unfortunate that the SI unit system has a scaled base unit, but we can't change that.

Still, I think it might be better for unlib and the sanity of its users to:
- use _gram_ for the mass dimension exponent
- define the base unit kilogram as `template <typename V> using kilogram = quantity<mass              , kilo_scaling, V>;
- define all derived units in terms of the SI base units (or other derived units)

Author: OK, you are right on this. I will make `gram` the unscaled mass quantity again. Since all units are now defined in terms of quantities, using `kilogram` in the definition of, e.g. Newton, will make sure the derivations are correct nevertheless.

However... You just called `quantity<mass , kilo_scaling, V>` a "unit" and we just had this in verbal discussion, too. If so, then maybe "quantity" really _is_ indeed a misnomer? * sigh * (That's not without precedence. [Nic Holthaus' popular unit library](https://github.com/nholthaus/units) does this.) OTOH, in the unlib, a "unit" actually would be a combination of a quantity's dimension and scale. ("Mass" and "kilo" makes "kilogram".) For that, however, there is no actual type in the library. So this is a unit library that doesn't have units. `:-x` It would, of course, be easy to introduce another abstraction `unit`, consisting of dimension and a scale, and have `quantity` take a `unit`, rather than a dimension and a scale. I considered this an unnecessary, and purely abstract, layer, but if it helps us to talk about the types and objects instantiated, maybe it would help?

```

## Values

> Any integer or floating point type can be used as a quantity's value type. However, the literal operators provided by the library (in namespace `unlib::literals`) will return `unlib::literals::integer_value_type` (currently an alias for `std::int32_t`) and `unlib::literals::floatpt_value_type` (currently an alias for **`double`**). If you need to use `std::int64_t` or **`long double`**, but also want to use literals, you have seeveral choices:

Reviewer: ```suggestion
Any integer or floating point type can be used as a quantity's value type. However, the literal operators provided by the library (in namespace `unlib::literals`) will return `unlib::literals::integer_value_type` (currently an alias for `std::int32_t`) and `unlib::literals::floatpt_value_type` (currently an alias for **`double`**). If you need to use `std::int64_t` or **`long double`**, but also want to use literals, you have several choices:
```

Author: 👍🏻

(_Note: Using tags, the library already provides the three incompatible quantities `watt` for active power, `var` for reactive power, and `voltampere` for apparent power. Those are, again, templates, which can be parametrized with a value type.)_

Two quantities can be added and subtracted if both have the same tag (which can be `no_tag`). They can be multiplied and divided if

 1. either operand has the `no_tag` tag or 
 1. both operands have the same tag (which can be `no_tag`).

When multiplying and dividing tagged quantities, the library keeps track of the tag's "exponents" (how many times quantities of the same tag have been multiplied or divided with). 

Therefore you can, for example, multiply _reactive power_ with _time_, resulting in _reactive energy_. This divided by time results again in reactive power. If it is divided by reactive power, the result will be (untagged) time. You can also square _reactive power_, resulting in a quantity with a _reactive_ tag exponent of _2/1_. If you then divide this by _reactive power_, the result will be _reactive power_ again.)

> Note that tags do not reflect all properties of their engineering counterparts. For example, dividing _active power_ by _voltage_, which is untagged, will result in the resulting quantity of _current_ to be tagged as _reactive_, which very likely won't make much sense. In these cases you will have to use a [tag_cast](#conversions) to make the library submit to your application domain's rules. 

Reviewer: ```suggestion
Note that tags do not reflect all properties of their engineering counterparts. For example, dividing _reactive power_ by _voltage_, which is untagged, will result in the resulting quantity of _current_ to be tagged as _reactive_, which very likely won't make much sense. In these cases you will have to use a [tag_cast](#conversions) to make the library submit to your application domain's rules. 
```

Author: 👍🏻
